storage/aptosdb/src/state_kv_db.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    db_metadata::{DbMetadataKey, DbMetadataSchema, DbMetadataValue},
+    db_options::{gen_state_kv_cfds, state_kv_db_column_families},
+    NUM_STATE_SHARDS,
+};
+use anyhow::Result;
+use aptos_config::config::{RocksdbConfig, RocksdbConfigs};
+use aptos_logger::prelude::info;
+use aptos_rocksdb_options::gen_rocksdb_options;
+use aptos_schemadb::{SchemaBatch, DB};
+use aptos_types::transaction::Version;
+use arr_macro::arr;
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+pub const STATE_KV_DB_FOLDER_NAME: &str = "state_kv_db";
+pub const STATE_KV_METADATA_DB_NAME: &str = "state_kv_metadata_db";
+
+pub struct StateKvDb {
+    state_kv_metadata_db: Arc<DB>,
+    _state_kv_db_shards: [Arc<DB>; NUM_STATE_SHARDS],
+}
+
+impl StateKvDb {
+    // TODO(grao): Support more flexible path to make it easier for people to put different shards
+    // on different disks.
+    pub(crate) fn new<P: AsRef<Path>>(
+        db_root_path: P,
+        rocksdb_configs: RocksdbConfigs,
+        readonly: bool,
+        ledger_db: Arc<DB>,
+    ) -> Result<Self> {
+        if !rocksdb_configs.use_state_kv_db {
+            info!("State K/V DB is not enabled!");
+            return Ok(Self {
+                state_kv_metadata_db: Arc::clone(&ledger_db),
+                _state_kv_db_shards: arr![Arc::clone(&ledger_db); 256],
+            });
+        }
+
+        let state_kv_db_config = rocksdb_configs.state_kv_db_config;
+        let state_kv_metadata_db_path = db_root_path
+            .as_ref()
+            .join(STATE_KV_DB_FOLDER_NAME)
+            .join("metadata");
+
+        let state_kv_metadata_db = Arc::new(Self::open_db(
+            state_kv_metadata_db_path.clone(),
+            STATE_KV_METADATA_DB_NAME,
+            &state_kv_db_config,
+            readonly,
+        )?);
+
+        info!(
+            state_kv_metadata_db_path = state_kv_metadata_db_path,
+            "Opened state kv metadata db!"
+        );
+
+        // TODO(grao): Support sharding here.
+        let sharding = false;
+        let _state_kv_db_shards = {
+            if sharding {
+                let mut shard_id: usize = 0;
+                arr![{
+                    let db = Self::open_shard(db_root_path.as_ref(), shard_id as u8, &state_kv_db_config, readonly)?;
+                    shard_id += 1;
+                    Arc::new(db)
+                }; 256]
+            } else {
+                arr![Arc::clone(&state_kv_metadata_db); 256]
```

Comment:
why do we want to use state_kv_metadata_db for kv_db_shards when sharding is not on?